python/google/protobuf/importer.py

```diff
@@ -0,0 +1,237 @@
+          _sym_db.RegisterMessage(message_type)
+          setattr(module, name, message_type)
+
+    class ProtoFinder(importlib.abc.MetaPathFinder):
+      """MetaPathFinder that locates .proto files."""
+
+      def find_spec(self, fullname, path, target=None):
+        """Determines whether a .proto file can be loaded."""
+        del path
+        del target
+        filepath = _module_name_to_proto_file(fullname)
+        for search_path in sys.path:
+          try:
+            prospective_path = os.path.join(search_path, filepath)
+            os.stat(prospective_path)
+          except (FileNotFoundError, NotADirectoryError):
+            continue
+          else:
+            return importlib.machinery.ModuleSpec(
+                fullname, ProtoLoader(fullname, filepath))
+
+    def protos(protobuf_path, include_paths=None):  # pylint: disable=g-function-redefined, function-redefined
+      """Loads a module from a .proto file on disk.
+
+        This function is idempotent. Invoking it a second time will simply
+        return a module that has already been loaded. If the desired proto has
+        already been loaded from a "_pb2.py" file, this function will simply
+        return that module.
+
+        This function guarantees pointer equality for all types in the
+        dependency tree of an imported type. For example, suppose we have three
+        .proto files: A.proto, B.proto, and C.proto with the following contents:
+
+        A.proto:
+        ```
+        import "C.proto"
+
+        message AMessage {
+            CMessage c = 1;
+        }
+        ```
+
+        B.proto:
+        ```
+        import "C.proto"
+
+        message BMessage {
+            CMessage c = 1;
+        }
+        ```
+
+        C.proto:
+        ```
+        message CMessage {
+            int64 foo = 1;
+        }
+        ```
+
+        Then we guarantee that the following assertion will hold regardless of
+        the combination of generated code and runtime import used to load the
+        three modules:
+
+        ```
+        assert AMessage().c.__class__ is BMessage().c.__class__
+        ```
+
+      Args:
+        protobuf_path: A string representing the file path to the .proto file
+          on disk. (e.g. "google/protobuf/any.proto"). By default, all the
+          paths on sys.paths are searched for this file.
+        include_paths: An optional sequence of strings representing paths on
+          which to search for the .proto file to search in addition to the
+          entries in sys.paths.
+
+      Returns:
+        A module object identical to one that would be generated by protoc
+        and loaded with an `import foo_pb2` statement.
+      """
+      with _augmented_syspath(include_paths):
+        module_name = _proto_file_to_module_name(protobuf_path)
+        module = importlib.import_module(module_name)
+        return module
+
+    sys.meta_path.extend([ProtoFinder()])
```

Comment:
During development, you may actually want the ProtoFinder to go ahead of normal *.py resolution (so that a *.proto file with a newer mtime takes precedence over the locally commited generated *.py file).

Reply:
Conversely, in production, you want "_pb2.py" files to take precedence over ".proto" files. Perhaps a feature flag/environment variable that allows you two choose between the two behaviors? I'd prefer an argument to the function, but Finders unfortunately have to be per-process globals.